pkg/logs/internal/decoder/auto_multiline_detection/token_graph.go

```diff
@@ -0,0 +1,120 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package automultilinedetection contains auto multiline detection and aggregation logic.
+package automultilinedetection
+
+// TokenGraph is a directed cyclic graph of tokens that model the relationship between any two tokens.
+// It is used to calculate the probability of an unknown sequence of tokens being represented by the graph.
+type TokenGraph struct {
+	adjacencies        [][]bool
+	minimumTokenLength int
+}
+
+// MatchContext is the context of a match.
+type MatchContext struct {
+	probability float64
+	// start and end are the indices of the token subsequence that produced the highest probability.
+	start int
+	end   int
+}
+
+// NewTokenGraph returns a new TokenGraph.
+func NewTokenGraph(minimumTokenLength int, inputData [][]Token) *TokenGraph {
+	g := &TokenGraph{
+		adjacencies:        make([][]bool, end),
+		minimumTokenLength: minimumTokenLength,
+	}
+	for _, tokens := range inputData {
+		g.add(tokens)
+	}
+	return g
+}
+
+// add adds a sequence of tokens to the graph.
+func (m *TokenGraph) add(tokens []Token) {
+	lastToken := tokens[0]
+	for _, token := range tokens[1:] {
+		if m.adjacencies[lastToken] == nil {
+			m.adjacencies[lastToken] = make([]bool, end)
+		}
+		m.adjacencies[lastToken][token] = true
+		lastToken = token
+	}
+}
+
+// MatchProbability returns the probability of a sequence of tokens being represented by the graph.
+func (m *TokenGraph) MatchProbability(tokens []Token) MatchContext {
+	if len(tokens) < 2 {
```

Comment:
```suggestion
	if len(tokens) < m.minimumTokenLength {
```

Since shorter `tokens` would never produce a long enough sequence of matches to satisfy condition on line 73.